File: backend/apps/authentication/validators/ruc_validator.py
"""
RucValidator — Chain of Responsibility node validating the RUC field (SRP).
"""

from __future__ import annotations

from core.base.base_validator import BaseValidator, ValidationResult

RUC_LENGTH = 13


class RucValidator(BaseValidator):
    def validate(self, data: dict) -> ValidationResult:
        ruc = (data.get("ruc") or "").strip()
        if not ruc:
            return ValidationResult(
                is_valid=False,
                errors=["El RUC es obligatorio."],
                field_name="ruc",
            )
        if not (ruc.isdigit() and len(ruc) == RUC_LENGTH):
            return ValidationResult(
                is_valid=False,
                errors=["El RUC debe tener 13 dígitos numéricos."],
                field_name="ruc",
            )
        return ValidationResult(is_valid=True)

File: frontend/src/modules/auth/validators/RucValidator.ts
/**
 * RucValidator — Chain of Responsibility node for RUC validation (FE).
 * Policy: exactly 13 numeric digits (Ecuadorian RUC).
 * Extends BaseValidator (core). SOLID: SRP·OCP·LSP.
 */

import { BaseValidator, type ValidationResult } from '../../../core/base/BaseValidator'

const RUC_RE = /^\d{13}$/

export class RucValidator extends BaseValidator {
  validate(data: unknown): ValidationResult {
    const ruc = String((data as { ruc?: string })?.ruc ?? '').trim()
    if (!ruc) {
      return { isValid: false, errors: ['El RUC es obligatorio.'], field: 'ruc' }
    }
    if (!RUC_RE.test(ruc)) {
      return {
        isValid: false,
        errors: ['El RUC debe tener exactamente 13 dígitos numéricos.'],
        field: 'ruc',
      }
    }
    return { isValid: true, errors: [], field: 'ruc' }
  }
}
